python/ml4ir/applications/ranking/model/layers/masking.py

```diff
@@ -159,6 +165,31 @@ def __init__(self,
         self.mask_rate = mask_rate
         self.mask_at_inference = mask_at_inference
         self.requires_mask = requires_mask
+        self.fixed_masks = None
+        self.current_mask_index = -1
+
+    def apply_stochastic_mask(self, inputs, training):
+        batch_size, sequence_len, feature_dim = tf.shape(inputs)[0], tf.shape(inputs)[1], tf.shape(inputs)[2]
+        if self.mask_at_inference or training:
+            mask = tf.cast(tf.math.greater(tf.random.uniform([batch_size, feature_dim]), self.mask_rate),
+                           dtype=tf.float32)
+            mask = tf.expand_dims(mask, axis=1)
+            mask = tf.tile(mask, [1, sequence_len, 1])
+            return tf.multiply(inputs, mask)
+        else:
+            return inputs
+
+    def apply_fixed_mask(self, inputs, training):
+        batch_size, sequence_len, feature_dim = tf.shape(inputs)[0], tf.shape(inputs)[1], tf.shape(inputs)[2]
+        if self.fixed_masks is None:
+            self.fixed_masks = list(itertools.product([0, 1], repeat=feature_dim))
+            QueryFeatureMask.masking_config[MonteCarloInferenceKey.FIXED_MASK_COUNT] = len(self.fixed_masks)
+
+        current_mask = tf.expand_dims(tf.expand_dims(self.fixed_masks[self.current_mask_index], 0), 0)
+        current_mask = tf.tile(current_mask, [batch_size, sequence_len, 1])
+        self.current_mask_index += 1
+        self.current_mask_index %= QueryFeatureMask.masking_config[MonteCarloInferenceKey.FIXED_MASK_COUNT]
```

Comment:
@mohazahran so we just have to make sure to call from scorer the exact same number of times as mask count, correct?

Reply:
Yes, but this is done automatically in the monteCarloScorer.